MSBuild Task for Semantic Versioning & Tests

JohnTortugo opened the pull request:
Implementation of generation of Versioning Strings as [described in the plan.](https://github.com/dotnet/arcade/issues/134)

- [x] Support for SemVer. 2.0
- [x] Support for SemVer. 1.0
- [x] Store computed information on a file to prevent unecessary re-computation.
- [x] Migrate task for extracting date+revision from OfficialBuildId
- [x] Add basic testing for Targets and Tasks
- [x] Documentation with usage examples

Suggestions are welcome.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Microsoft.DotNet.Build.Tasks.SemVer
+{
+    public class SemVer : Task
```
I think we need a better name then SemVer.

JohnTortugo replied:
I am all open for better names.

mmitche replied:
Yeah, I think: `Versioning` is better.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,55 @@
+            }
+
+            // So far we know that these first four fields aren't empty
+            Version = $"{Major}.{Minor}.{Patch}-{Prerelease}.{ShortDate:00000}.{Builds}+{ShortSHA}";
```
Is this only to concat the versions? Why do we need a task for that over just a targets file?

JohnTortugo replied:
The requirements for this weren't clear yet. I was waiting feedback from @mmitche to continue the implementation. I just pushed new changes. I think if we don't end up with something too complex, transforming this Task to an inline task seems reasonable.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,63 @@
+    {
+        private Dictionary<string, string> formats = new Dictionary<string, string> {
+                                    {"dev", "{major}.{minor}.{patch}-{prerelease}.{shortdate}.{builds}+{shortsha}"},
+                                    {"stable", "{major}.{minor}.{patch}-{prerelease}"},
```
I'd make this "stable-prerelease" since technically both stable and final are 'stabilized'

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Microsoft.DotNet.Build.Tasks.Versioning.csproj:
```diff
@@ -0,0 +1,29 @@
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetFxTfm)'">
+    <Reference Include="Microsoft.Build.Tasks.v4.0" />
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
```
Why reference the ancient MSBuild here? Do you have a requirement of building on machines with MSBuild 4?

JohnTortugo replied:
Will remove that in next commit. Thanks.

mmitche replied:
@JohnTortugo I think this needs to be removed.

JohnTortugo replied:
Thanks Matt. Removed.

JohnTortugo replied:
Hi @mmitche , I had copied this configuration from the Task.IO package. That being said I've a few questions: 
- Do we need to target multiple frameworks on the Versioning target? 
- What is this $(NextFxTfm) target? 
- If we need to target $(NetFxTfm), which Microsoft.Build package version should I use? Thanks.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Versioning.cs:
```diff
@@ -0,0 +1,87 @@
+            VersionString = VersionString.Replace("{builds}", Builds.ToString());
+            VersionString = VersionString.Replace("{shortsha}", ShortSHA);
+
+            return true;
```
Prefer `return !Log.HasLoggedErrors` for all exits in MSBuild tasks. This helps guard against the case where the task fails but doesn't log an error and is super hard to track down.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    public class SemVer : Task
+    {
+        private Dictionary<string, string> formats = new Dictionary<string, string> {
```
This can be initialized with a `CaseInsensitiveComparer`, avoiding the ToLower below

mmitche commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,63 @@
+                VersionString = formats[VersionString];
+            }
+
+            if ((VersionString.Contains("{major}") && Major == 0) ||
```
I think this is a valid version. Technically for SemVer2, you start with pre-release versions that are 0.X.Y-blah, then rename 1.0.0 for the initial release.

JohnTortugo replied:
Thanks. Will fix that.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.SemVer/SemVer.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            VersionString = FormatString.ToLower();
+
+            if (formats.ContainsKey(VersionString))
```
Does this allow for proper overriding?  I think what might make more sense is to allow an input itemgroup that has the dev/stable/final format strings, remove FormatString, and then rename VersionString to something indicating it means the state.  Maybe `VersionState`.  That means the logic is like:

```
if OverrideFormatStrings.ContainsKey(versionState) then use override format string
else if DefaultFormatStrings.ContainsKey(versionState) the use default format string
else error

JohnTortugo commented:
Talked with @weshaggard ; I am going to work on the changes he suggested.

JohnTortugo commented:
Hi guys, I've applied the changes. PTAL

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+===============================
+
+Task package which handles generation of version strings. 
+The format that we use is described here: Documentation/Versioning.md
```
Can you make this a relative link?

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+ShortSha        | string      | SHA of the repo last commit.
+FormatName      | string      | The name of the format string you want to use. Defaults are "dev", "stable" and "final".
+FormatStrings   | ItemGroup   | Your custom format strings.
+ExitCode        | integer     | **Output** Indicate whether an error happened (1) or not (0).
```
Why is the ExitCode needed?  Typically the target would just call the Error target on any issue.

JohnTortugo replied:
Thanks. Changed it to <Error/>

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Versioning.props:
```diff
@@ -0,0 +1,29 @@
+    <PropertyGroup>
+      <VersionString Condition="'%(FormatStrings.Identity)'=='$(FormatName)'">%(FormatStrings.Format)</VersionString>
+
+      <ExitCode Condition="'$(VersionString)'!=''">0</ExitCode>
```
These can be replaced with an <Error> element.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Versioning.props:
```diff
@@ -0,0 +1,29 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Target Name="Versioning">
+
```
Should the target include a call to the date generation task if the input date is empty?

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+      <Format>$(SomethingCustom) $(major).$(minor).$(patch)-$(prerelease).$(shortdate).$(builds)+$(shortsha)</Format>
+    </FormatStrings>
+
+    <FormatStrings Include="stable">
```
this isn't what I consider stable. I consider it a "final prerelease"

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+      <Format>$(SomethingCustom) $(major).$(minor).$(patch)-$(prerelease)</Format>
+    </FormatStrings>
+
+    <FormatStrings Include="final">
```
this is what I consider stable.

JohnTortugo replied:
@mmitche suggested those names/formats to me. Is there a list of names and formats for this? 

Also, this is just some default names/formats, the user of the Versioning task will be able to override or add custom ones.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+    </FormatStrings>
+  </ItemGroup>
+  
+  <Target Name="TestIt" DependsOnTargets="Versioning">
```
Do you intend to keep this "TestIt" target?

JohnTortugo replied:
It was intended just as an example in the documentation. However, I might in fact use it to implement some tests for the versioning task..; I am open to sugestions..

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,149 @@
+
+namespace Microsoft.DotNet.Build.Tasks.Versioning
+{
+    public sealed class GenerateVersioningDate : Task
```
Perhaps I'm missing it but where is this task called?

JohnTortugo replied:
It'll be called when the user don't inform a date to the versioning task.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Versioning.props:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Target Name="Versioning">
```
What is the intend of this target exactly? Is it simply to fill in the VersionString? If so where does that get used?

JohnTortugo replied:
As far as I know that was the main goal. I understood that it was just to generate strings based on the plan described here: https://github.com/dotnet/arcade/pull/108

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Versioning.props:
```diff
@@ -0,0 +1,29 @@
+    </ItemGroup>
+    
+    <PropertyGroup>
+      <VersionString Condition="'%(FormatStrings.Identity)'=='$(FormatName)'">%(FormatStrings.Format)</VersionString>
```
Should this be VersionFormatString? Where doe you expect it to be used and have the values substituted?

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+    <Minor>2</Minor>
+    <Patch>3</Patch>
+    <Prerelease>beta1</Prerelease>
+    <ShortDate>3111985</ShortDate>
```
How are repo's expected to setup all these values? Surely they aren't all going to be hard-coded so it seems like there should be more of these moved into the common props/targets with the versioning tasks. I would also expect this task library to handle the work of caching the values that have to gathered during the build (shortdate, shortsha etc), such that projects don't have to worry about it and only need to be sure they import the props/targets file from this versioning package.

JohnTortugo replied:
What I understood from @mmitche is that we would create the package in a way that users would be able to customize it by adding a .props file or passing the properties as command line parameters.

Regarding the caching, do you mean like this: https://github.com/dotnet/buildtools/blob/d21b572f15e2dcf7a67a43c77fb0b5a5e7d9aaf4/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L410 ?

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,62 @@
+    </FormatStrings>
+
+    <FormatStrings Include="stable">
+      <Format>$(SomethingCustom) $(major).$(minor).$(patch)-$(prerelease)</Format>
```
We should append -final to the final prerelease version to avoid version precedence issues.

JohnTortugo commented:
> Is it possible to add a test for the new target?

I think this is important to have. Can you guys suggest what is the best approach to test&automate the Target itself?

> I would also expect this task library to handle the work of caching 

Will start working on that.

JohnTortugo commented:
Can you guys please comment on:
- exactly what should be cached and,
- where should this file be created?

I am using the same file structure created [here](https://github.com/dotnet/buildtools/blob/d21b572f15e2dcf7a67a43c77fb0b5a5e7d9aaf4/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L410) ; Are we going to keep all those fields? There was no mention for the BuildByString parameter in the versioning plan; should I incorporate it as well?

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Util/Microsoft.DotNet.Build.Tasks.Util.csproj:
```diff
@@ -0,0 +1,28 @@
+
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetFxTfm)'">
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
```
Is this needed for the netfx TFM?

JohnTortugo replied:
Yeap, exactly.

rainersigwald replied:
Why can't you just reference the NuGet package both ways?

JohnTortugo replied:
I am not sure exactly what you mean :-/

joperezr replied:
I believe what he means is that instead of having these two references, you could probably just have the `<PackageReference Include="Microsoft.Build.Framework" Version="15.1.548" />` in both netcoreapp2.0 and netfxtfm and that should provide the right assets for each TFM instead of referencing the platform version like you are doing here.

---
In reply to: [187209459](https://github.com/dotnet/arcade/pull/137#discussion_r187209459) [](ancestors = 187209459)

JohnTortugo replied:
Thanks @joperezr . It'll go up in the next commit.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,65 @@
+    <Major>1</Major>
+    <Minor>2</Minor>
+    <Patch>3</Patch>
+    <FormatName>stable</FormatName>
```
I think the pre-release label goes here too

JohnTortugo replied:
This was just an example. The PreRelease field isn't used in the stable format string.

weshaggard replied:
It should become what people are expected to copy/paste into their repo so we should make it as complete as possible.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.props:
```diff
@@ -0,0 +1,9 @@
+    <Major>1</Major>
+    <Minor>0</Minor>
+    <Patch>0</Patch>
+  </PropertyGroup>
```
Prerelease label should go here too.  `preview1` seems like a reasonable default.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+    <_MicrosoftDotNetBuildTasksVersioningTaskAssembly Condition="'$(MSBuildRuntimeType)'!='Core'">$(MSBuildThisFileDirectory)net45\Microsoft.DotNet.Build.Tasks.Versioning.dll</_MicrosoftDotNetBuildTasksVersioningTaskAssembly>
+  </PropertyGroup>
+
+  <Target Name="Versioning">
```
I think this target should be call GenerateVersions.

weshaggard replied:
I would expect it to compute the Version, FileVersion, InformationalVersion properties that the SDK uses. See https://github.com/dotnet/sdk/blob/2e6c5933e0553c02a8b6bc97206ceb1abd10da9c/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L166 for the properties that the SDK uses. This target should likely be a BeforeTarget of GetAssemblyVersion and/or GenerateAssemblyInfo.

weshaggard replied:
Have a look at https://github.com/dotnet/roslyn-tools/blob/cf50106f72f413a63a2d22087cd5c695cd480943/sdks/RepoToolset/tools/Version.targets which does something very similar.

JohnTortugo replied:
Questions:

- Are we going to need to differentiate CIBuild, OfficialBuild, "Local build"? I ask that because it seems that InformationalVersion, for instance, depends on the type of the build.
- Version == VersionString?
- FileVersion == "$(Major).$(Minor).$(Patch).$(Build,5)" ?

Cc'ing @mmitche

weshaggard replied:
Yes I do think we will want extra data in the information version, see https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L380 for some examples.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+    <!-- Delete old BuildVersion.props files -->
+    <Delete Files="$(BuildVersionFile)" TreatErrorsAsWarnings="true"/>
+    
+    <WriteLinesToFile
```
The generation of this props file should be in its own target so that we can control whether or not we runs as we don't want it to run all the time for every individual project.

JohnTortugo replied:
Got it. Will push on next commit. Thanks.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+      Create a cache file so that other projects don't have to recompute these info on every build.
+    -->
+    <PropertyGroup>
+      <BuildVersionFile>$(MSBuildThisFileDirectory)BuildVersion.props</BuildVersionFile>
```
We shouldn't be putting this file in the same directory as this targets file. It should be going into an intermediate directory.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+    -->
+    <ItemGroup>
+      <FormatStrings Include="dev">
+        <Format>$(major).$(minor).$(patch)-$(prerelease).$(shortdate).$(builds)$(sha)</Format>
```
These properties should match how they are defined, including case.

mmitche replied:
@JohnTortugo I see what wes is going for here.  The point is to avoid the format strings itemgroup so that all possible version strings don't appear in a detailed/diagnostic log, even though only one gets used.  To achieve that, the `VersionString` property is just calculated iteratively.  e.g.

```
<PropertyGroup>
  <VersionString>$(Major).$(Minor).$(Patch)</VersionString>
 <VersionString Condition="'$(FormatName)' !='stable'>$(VersionString)-$(PreReleaseLabel)</VersionString>
  <VersionString Condition="'$(FormatName)' =='dev' and '$(IncludeDate)' == 'true'>$(VersionString).$(GeneratedDate).$(GeneratedRevision)</VersionString>
...
</PropertyGroup>
```

JohnTortugo replied:
Thanks @mmitche . I got your point. The code also looks much cleaner that way.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+    </PropertyGroup>
+
+    <!-- Delete old BuildVersion.props files -->
+    <Delete Files="$(BuildVersionFile)" TreatErrorsAsWarnings="true"/>
```
Why delete if you're going to overwrite it in the next task?

JohnTortugo replied:
Thanks. Removed.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+<!-- This file was automatically generated by the MSBuild Versioning target. -->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VersionPropsImported>true</VersionPropsImported>
```
The generated properties should be written to this file, ShortDate, Builds, and Sha, matching the casing and names, such that importing this generated file before running the target wouldn't run GenerateVersioningDate or the future sha retrieval task again.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VersionPropsImported>true</VersionPropsImported>
+    <BuildNumberMajor Condition="'%24(BuildNumberMajor)' == ''">$(major)</BuildNumberMajor>
```
Anything that should be static in the source (pre-release label, major, minor, and patch) doesn't need to be written to this file.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+      <VersionFileContent>
+<![CDATA[<?xml version="1.0" encoding="utf-8"?>
+<!-- This file was automatically generated by the MSBuild Versioning target. -->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
In modern MSBuild (15+) you don't need the namespace, so you can keep this cleaner. Also don't need the xml declaration.

JohnTortugo replied:
Removing.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+    -->
+    <ItemGroup>
+      <FormatStrings Include="dev">
+        <Format>$(major).$(minor).$(patch)-$(prerelease).$(shortdate).$(builds)$(sha)</Format>
```
Can you match the casing with the input properties?

JohnTortugo replied:
Sure. Will do that.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+<!-- This file was automatically generated by the MSBuild Versioning target. -->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VersionPropsImported>true</VersionPropsImported>
```
How and where will this file be imported? I'm worried that you're going to run into trouble if it's updated within a build.

JohnTortugo replied:
I thought that projects would import this auto-generated file directly as part of their build process. But after the last comments it seems that they will actually import the Versioning Task and this cache file will actually be used by the versioning task itself. @mmitche / @weshaggard , is my understanding correct about this?

rainersigwald replied:
Ok, neither of those approaches is going to work right, unfortunately. Feel free to schedule some time to chat with me about it if you like. The main issue is https://github.com/Microsoft/msbuild/issues/1185 . . . you'll get stale data if you update the "cache" file within the build.

weshaggard replied:
For a full build we should generate this file once and then import it but it shouldn't change during the build itself. 

This is how we do it today in BuildTools https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/BuildVersion.targets#L10. If it exists we import it otherwise we generate it.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,160 @@
+        /// The Revision number that will be produced from the BuildNumber.
+        /// </summary>
+        [Output]
+        public string GeneratedDate { get; set; }
```
These should be named what they are representing:
- GeneratedRevision - (build revision)
- GeneratedDate (5 digit date)

JohnTortugo replied:
OK. Renamed.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+      This task will also try to extract the date from the BuildId.
+    -->
+    <GenerateVersioningDate Condition="'$(IncludeDate)'=='true' and '$(ShortDate)'==''" SeedDate="$(VersionSeedDate)" OfficialBuildId="$(OfficialBuildId)" ComparisonDate="$(VersionComparisonDate)" Padding="$(VersionPadding)">
+      <Output TaskParameter="GeneratedVersion" PropertyName="GeneratedVersion"  />
```
The GeneratedVersion feeds into Builds

JohnTortugo replied:
Thanks. I was going to ask about that. Also, what about the condition for calling GenerateVersioningDate. Does OfficialBuildId have priority over ShortDate and Builds, right? I mean, whenever OfficialBuildId != "", ShortDate and Builds will be recomputed, even if they are not empty.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,160 @@
+        /// <summary>
+        /// Optional parameter that sets the Padding for the version number. Must be 5 or bigger.
+        /// </summary>
+        public int Padding { get; set; }
```
We are going to need an option that turns off padding, see comment https://github.com/dotnet/arcade/pull/108#discussion_r186895322, as semver 2 doesn't like padding for numbers.

JohnTortugo replied:
Isn't that rule 9? " ... Numeric identifiers MUST NOT include leading zeroes ...." ?
Should it still be optional since the spec say it **must not** include the leading?

weshaggard replied:
Your right that is rule 9. However we still need this to be optional because we do need the leading zero padding to support correct sorting in semver 1 scenarios.

JohnTortugo replied:
Got it, thanks.

weshaggard commented:
>I think this is important to have. Can you guys suggest what is the best approach to test&automate the Target itself?

For testing I think you could create a simple library project that references this package/project and uses the versioning scheme. Not sure if a ProjectReference will get the target file correctly imported but you could statistically import it to verify.

JohnTortugo commented:
> For testing I think you could create a simple library project that references this package/project and uses the versioning scheme. Not sure if a ProjectReference will get the target file correctly imported but you could statistically import it to verify.

Thanks! I was working on something similar to this. I searched a bit and it's possible to call a target from a C# piece of code. I am thinking to plug that together with xUnit and that way be able to automate & test many different inputs. Please, let me know what you think.

JohnTortugo commented:
Added some changes:

- Removed Format Strings Itemgroup; Dropped this approach.
- Fixed property cases and default values.
- Renamed the versioning task.
- Added new target to create cache file.
- Generating cache file only when it doesn't exist.
- Importing cache file
- Support for semantic versioning 1.0.
- Support for padding fields: **Still work in progress.**
- Cache file content: **Still work in progress.**

JohnTortugo commented:
**Update and request for comments:** I tried a few options today to automate the test of the new target; basically, could not make it work nicely. I tried two options:

- I tried "programatically" (like [this](https://stackoverflow.com/a/42816251/973795)) calling the target from inside xUnit tests but things wen't crazy as it required setting some environment variables, specifying msbuild "installation" paths, etc, [something like this.](https://stackoverflow.com/a/43507772/973795) and still didn't work properly.
- I also tried to "automate" it using only MSBuild targets. Something like creating a Test target that calls the GenerateVersions target multiple times with different parameters. The problem with that was mostly that I wasn't able to call the same target multiple times..

I talked with @rainersigwald and he said that actually there is not a good way to automate this; even more in .NET core. He suggested that we can try to do something similar to what the [SDK team did](https://github.com/dotnet/sdk/blob/master/src/Tests/Microsoft.NET.Build.Tests/GivenThereAreDefaultItems.cs#L30-L62) or maybe just some scripting(?) wrapping calls to "dotnet msbuild" passing some Test.xml file.

I think the option of creating a few TestCase01.xml, TestCase02.xml, etc files and them calling "dotnet msbuild" is interesting, because: 1) it's simple; 2) we have quite some control of the test and the output; 3) Is quite portable across .net core and .net full?! 4) Do we need something more sophisticated?

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -3,89 +3,82 @@
   <PropertyGroup>
     <_MicrosoftDotNetBuildTasksVersioningTaskAssembly Condition="'$(MSBuildRuntimeType)'=='Core'">$(MSBuildThisFileDirectory)netcoreapp2.0\Microsoft.DotNet.Build.Tasks.Versioning.dll</_MicrosoftDotNetBuildTasksVersioningTaskAssembly>
     <_MicrosoftDotNetBuildTasksVersioningTaskAssembly Condition="'$(MSBuildRuntimeType)'!='Core'">$(MSBuildThisFileDirectory)net45\Microsoft.DotNet.Build.Tasks.Versioning.dll</_MicrosoftDotNetBuildTasksVersioningTaskAssembly>
+    <VersioningCacheFile>$(MSBuildThisFileDirectory)VersioningCache.props</VersioningCacheFile>
```
This still needs to be an intermediate directory of some sort.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,84 @@
+
+      <!-- Specific fields to "dev" format. Inclusion of Date and Sha fields are optional. -->
+      <TrailingPart Condition="'$(FormatName)'=='dev' and '$(IncludeDate)'=='true'">$(TrailingPart).$(ShortDate)</TrailingPart>
+      <TrailingPart Condition="'$(FormatName)'=='dev'">$(TrailingPart).$(Builds)</TrailingPart>
```
Isn't the Builds part captured in the ShortDate property?

weshaggard replied:
Nevermind I see it is separate. I think the property name "Builds" may be too general. Something like GeneratedRevisions like the task uses might be better.

JohnTortugo replied:
Renamed it to "Revision", though in the versioning plan it was named "Builds".

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,160 @@
+        public string GeneratedRevision { get; set; }
+
+        /// <summary>
+        /// The Revision number that will be produced from the BuildNumber.
```
These comments don't align with the property names.

weshaggard commented:
>I think the option of creating a few TestCase01.xml, TestCase02.xml, etc files and them calling "dotnet msbuild" is interesting, because: 1) it's simple; 2) we have quite some control of the test and the output; 3) Is quite portable across .net core and .net full?! 4) Do we need something more sophisticated?

Are those TestCase01.xml files csproj's? If so then that is similar to what I had in mind. Having some sample test projects that we call msbuild on and capture the output of the version in some way, or asserting it in the projects themselves by having them fail to build if the versions don't match what is expected.

JohnTortugo commented:
> Are those TestCase01.xml files csproj's? If so then that is similar to what I had in mind. Having some sample test projects that we call msbuild on and capture the output of the version in some way, or asserting it in the projects themselves by having them fail to build if the versions don't match what is expected.

Yes, that's exactly it.

JohnTortugo commented:
Hi @mmitche, @weshaggard / all; I believe we have sufficient material here to call this a minimal working version. Please, let me know what else I should work to get an approval for the merge :-)

joperezr commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,176 @@
+            }
+
+            // Leading zeros are only allowed in SemVer 1.0
+            if (IncludePadding)
```
Will padding ever be needed in case OfficialBuildId is passed in? I know it's not today, but is that something we want to support @weshaggard?

weshaggard replied:
If any repo's need to stick with SemVer 1 then yes they will need padding for there official builds. OfficialBuildId only contributes the date as well as the revision but both of those could still cause numbers less then the padding with and thus need padding.

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Versioning/README.md:
```diff
@@ -0,0 +1,134 @@
+# Microsoft.DotNet.Build.Tasks.Versioning
+
+Task package which handles generation of version strings. The versioning format that we use is described [here](../../Documentation/Versioning). See ["Task Packages"](../../Documentation/TaskPackages.md#usage) for guidance on installing this package.
```
I think that's a good point. I don't know if that was discussed already; I see that the other task packages on the project also don't have such automation, though.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -0,0 +1,176 @@
+
+                dateFromBuildId = dateFromBuildId.ToUniversalTime();
+                GeneratedShortDate = CreateShortDate(dateFromBuildId, ComparisonDate);
+                GeneratedRevision = match.Groups[2].Value;
```
You are not handling the padding for the revision. The OfficialBuildId will look something like `20180515-01` so you might need to remove or add padding for the revision.

JohnTortugo replied:
Actually I wasn't aware that we needed padding for the Revision. How much should it be? Two digits?

weshaggard replied:
It should be exactly what you parse when IncludePadding=true and it should have no leading zeros if IncludePadding=false.

mmitche commented:
Yeah, 2. Semver1 only.

From: Divino César <notifications@github.com>
Sent: Tuesday, May 15, 2018 2:58 PM
To: dotnet/arcade <arcade@noreply.github.com>
Cc: Matt Mitchell <mmitche@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/arcade] MSBuild Task for Semantic Versioning & Tests (#137)


@JohnTortugo commented on this pull request.

________________________________

In src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fpull%2F137%23discussion_r188449173&data=02%7C01%7Cmmitche%40microsoft.com%7C95a6f0be6e5c4966a89608d5baaedd0c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636620182555783776&sdata=P9hm9ttICt%2Fl8mxSg282HtjwzhB%2FytxTXcsGo0Eetis%3D&reserved=0>:

> +            Match match = regex.Match(buildId);

+

+            if (match.Success && match.Groups.Count > 2)

+            {

+                string dateFormat = "yyyyMMdd";

+                DateTime dateFromBuildId;

+

+                if (!DateTime.TryParseExact(match.Groups[1].Value, dateFormat, enUS, DateTimeStyles.AssumeLocal, out dateFromBuildId))

+                {

+                    Log.LogError("The OfficialBuildId doesn't follow the expected({0}.rr) format: '{1}'", dateFormat, match.Groups[1].Value);

+                    return false;

+                }

+

+                dateFromBuildId = dateFromBuildId.ToUniversalTime();

+                GeneratedShortDate = CreateShortDate(dateFromBuildId, ComparisonDate);

+                GeneratedRevision = match.Groups[2].Value;

Actually I wasn't aware that we needed padding for the Revision. How much should it be? Two digits?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fpull%2F137%23discussion_r188449173&data=02%7C01%7Cmmitche%40microsoft.com%7C95a6f0be6e5c4966a89608d5baaedd0c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636620182555793784&sdata=vYiC8Vcq1tot0qxgwRPCL9vPGzcBbtiwoiEua1sdajE%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIUisvh4gsE5jlcBuwVmvVyAskpT4woLks5ty09MgaJpZM4TnxL1&data=02%7C01%7Cmmitche%40microsoft.com%7C95a6f0be6e5c4966a89608d5baaedd0c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636620182555803793&sdata=TYUk%2Fx6Ae7LJMF4ummYnZ%2FbTy8h5sbTPifoWvss1h8g%3D&reserved=0>.

JohnTortugo commented:
Thanks. Pushed new update.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -60,7 +60,7 @@ public override bool Execute()
             if (string.IsNullOrEmpty(OfficialBuildId))
             {
-                GeneratedRevision = "0";
+                GeneratedRevision = (IncludePadding) ? "00" : "0";
```
For non-official builds it is ok to not be padded here, the fact is is just `0` actually helps identify it as a dev build.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersioningDate.cs:
```diff
@@ -123,6 +124,11 @@ private bool SetVersionAndRevisionFromBuildId(string buildId)
 
+                if (!IncludePadding)
+                {
+                    GeneratedRevision = int.Parse(GeneratedRevision).ToString();
```
Please use better error handling and do TryParse.

joperezr commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+  </Target>
+
+
+  <Target Name="CreateVersioningCacheFile" AfterTargets="GenerateVersions" Condition="'$(ShouldCreateVersionFileDuringBuild)'=='true'">
```
Wouldn't it be better if this is a DependsOnTargets?

joperezr commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+  <!-- If VersioningCacheFile exists already then import it; If it doesn't we are going to create it. -->
+  <Import Condition="Exists('$(VersioningCacheFile)')" Project="$(VersioningCacheFile)" />
+
+  <Target Name="GenerateVersions">
```
Where do you actually link this target into the build?

markwilkie replied:
And yes @joperezr , we intend to have arcade use this task.  Part of the point of arcade is to a) use stuff "first", and b) demonstrate a good pattern.

rainersigwald replied:
The way I interpreted the question was "What causes this target to be executed in an Arcade-using project?"

JohnTortugo replied:
What I had discussed with @mmitche was that "Arcade-using" projects would use this target as a depend-on of their own Build target. Did I got it right Matt?

rainersigwald replied:
That feels pretty heavyweight for consumers. Generally I'd recommend that a NuGet package or SDK that does something like this hook up to the build process using `BeforeTargets`.

mmitche replied:
@JohnTortugo That could be one way.  It could also be in the default targets that arcade provides and executes.  We should add an issue to add target dependency when it makes sense to do so.  I wouldn't block this PR on integration of this into the default arcade build process yet.

weshaggard replied:
If this package remains independent of the root arcade sdk package then I do believe we need to find an appropriate set of targets to chain this off of so folks only need to add the package reference and nothing else special is needed. However I agree with @mmitche that we don't need to block this PR on it but we should file an issue on figuring out the correct way to consume this package.

JohnTortugo replied:
Thank you for the feedback guys. I just created issue #176 for discussing this.

JohnTortugo commented:
Any objection on merging this PR once currently running checks pass?

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets:
```diff
@@ -0,0 +1,91 @@
+<!-- Copyright (c) .NET Foundation. All rights reserved. -->
```
Following the conventions of all our other tools packages this targets/props file should be called `Microsoft.DotNet.Build.Tasks.Versioning.props/targets`. In fact I think it might be needed to have nuget automatically reference theses files for projects that reference the package.

JohnTortugo replied:
I renamed the files and added the second part of your comment to the #176 issue.

alexperovich commented:
MSBuild/nuget will only automatically import `sdk\Sdk.props` and `sdk\Sdk.targets` from packages used in Sdk elements, and `build\<PackageName>.props` and `build\<PackageName>.targets` for packages referenced via PackageReference.

From: Wes Haggard <notifications@github.com>
Sent: Wednesday, May 16, 2018 2:12 PM
To: dotnet/arcade <arcade@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [dotnet/arcade] MSBuild Task for Semantic Versioning & Tests (#137)


@weshaggard commented on this pull request.

________________________________

In src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Versioning.targets<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fpull%2F137%23discussion_r188774366&data=02%7C01%7CAlex.Perovich%40microsoft.com%7Ce7406b67c8804338f37908d5bb71b808%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636621019449206149&sdata=W24Uoe%2Bm00ekD4FKEfN9W1Wuy20ycPIttlHJervTsVA%3D&reserved=0>:

> @@ -0,0 +1,91 @@

+<!-- Copyright (c) .NET Foundation. All rights reserved. -->

Following the conventions of all our other tools packages this targets/props file should be called Microsoft.DotNet.Build.Tasks.Versioning.props/targets. In fact I think it might be needed to have nuget automatically reference theses files for projects that reference the package.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fpull%2F137%23pullrequestreview-120822207&data=02%7C01%7CAlex.Perovich%40microsoft.com%7Ce7406b67c8804338f37908d5bb71b808%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636621019449216153&sdata=dJlYYwCX4bxaACApDSRdg8B%2BbW0r0GCTX%2FO5dj82a3k%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKE3IlGh-BdkVOzj8_EEM4a535A5Zr2Kks5tzJY3gaJpZM4TnxL1&data=02%7C01%7CAlex.Perovich%40microsoft.com%7Ce7406b67c8804338f37908d5bb71b808%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636621019449216153&sdata=ZW54ytxI9HiOeZDOVI7%2B2JbFKUi7K5cggqnIawCVZcY%3D&reserved=0>.
